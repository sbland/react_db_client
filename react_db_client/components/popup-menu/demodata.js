export const demoItems = [
  { uid: 'a', text: 'Item A', onClick: () => {} },
  { uid: 'b', text: 'Item B', onClick: () => {} },
  { uid: 'c', text: 'Item C', onClick: () => {} },
];

export const demoItemsEmpty = [];
