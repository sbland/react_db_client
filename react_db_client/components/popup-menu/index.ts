export { PopupMenu, RightClickWrapper } from './popup-menu';
