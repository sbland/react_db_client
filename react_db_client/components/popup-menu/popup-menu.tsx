import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import { getRoot } from '@react_db_client/helpers.get-root';
import { Uid } from '@react_db_client/constants.client-types';

import {
  PopupMenuListItemButtonStyle,
  PopupMenuListItemStyle,
  PopupMenuListStyle,
  PopupMenuOverlayStyle,
  PopupMenuRightClickWrapStyle,
  PopupMenuStyle,
  PopupMenuWrapStyle,
} from './styles';

export interface IItem {
  uid: Uid;
  label: string;
  onClick: () => void;
}
export interface IPopupMenuItemsProps {
  items: IItem[];
}

const PopupMenuItems = ({ items }: IPopupMenuItemsProps) => (
  <>
    {items.map((item) => (
      <PopupMenuListItemStyle className="popupMenu_menuItem" key={item.uid}>
        <PopupMenuListItemButtonStyle type="button" className="button-reset" onClick={item.onClick}>
          {item.label}
        </PopupMenuListItemButtonStyle>
      </PopupMenuListItemStyle>
    ))}
  </>
);

PopupMenuItems.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      onClick: PropTypes.func.isRequired,
    })
  ).isRequired,
};

export interface IPopupMenuProps {
  isOpenOverride?: boolean;
  onCloseCallback: (v: boolean) => void;
  position: { x: number; y: number };
  items: IItem[];
  popupRoot?: string | HTMLElement;
}

/**
 * Popup Menu
 *
 * @param {*} {
 *   isOpenOverride,
 *   items,
 * }
 * @returns
 */
export const PopupMenu = ({
  isOpenOverride,
  onCloseCallback,
  position,
  items,
  popupRoot,
}: IPopupMenuProps) => {
  const [isOpen, setIsOpen] = useState(isOpenOverride);

  const _popupRoot = getRoot(popupRoot);

  useEffect(() => {
    setIsOpen(isOpenOverride);
  }, [isOpenOverride]);

  const handleClose = (e) => {
    e.preventDefault();
    setIsOpen(false);
    onCloseCallback(false);
  };

  if (!_popupRoot) return <div>Missing Popup Root</div>;

  if (isOpen) {
    return ReactDOM.createPortal(
      <PopupMenuWrapStyle className="popupMenuWrap">
        <PopupMenuStyle
          className="popupMenu"
          style={{
            transform: `translate(${position.x}px, ${position.y}px)`,
          }}
        >
          <PopupMenuListStyle className="popupMenu_list">
            <PopupMenuItems items={items} />
          </PopupMenuListStyle>
        </PopupMenuStyle>
        {/* eslint-disable-next-line max-len */}
        {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */}
        <PopupMenuOverlayStyle
          className="popupMenu_overlay"
          onClick={handleClose}
          onContextMenu={handleClose}
        />
      </PopupMenuWrapStyle>,
      _popupRoot
    );
  }
  return <></>;
};

PopupMenu.propTypes = {
  isOpenOverride: PropTypes.bool,
  onCloseCallback: PropTypes.func.isRequired,
  position: PropTypes.shape({
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
  }).isRequired,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      onClick: PropTypes.func.isRequired,
    })
  ).isRequired,
  popupRoot: PropTypes.any,
};

PopupMenu.defaultProps = {
  isOpenOverride: false,
};

export interface IRightClickWrapperProps {
  children: React.ReactElement;
  items: IItem[];
  popupRoot?: string | HTMLElement;
}

export const RightClickWrapper = ({ children, items, popupRoot }: IRightClickWrapperProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const [pos, setPos] = useState({ x: 0, y: 0 });
  const handleContextMenu = (e) => {
    e.preventDefault();
    setIsOpen(true);
    setPos({ x: e.clientX, y: e.clientY });
  };

  return (
    <>
      {children?.type === 'div' ? (
        React.cloneElement(children, { onContextMenu: handleContextMenu })
      ) : (
        // : React.cloneElement(children, { ref: childRef }, [
        //   <div
        //     className="popupMenu_rightClickWrapper"
        //     onContextMenu={handleContextMenu}
        //     style={{
        //       width: '100%',
        //       height: '100%',
        //     }}
        //   />,
        // ])
        <PopupMenuRightClickWrapStyle
          style={{ width: '100%', height: '100%' }}
          className={`popupMenu_righClickWrap ${isOpen && 'active'}`}
          onContextMenu={handleContextMenu}
        >
          {children}
        </PopupMenuRightClickWrapStyle>
      )}
      <PopupMenu
        isOpenOverride={isOpen}
        onCloseCallback={() => setIsOpen(false)}
        items={items}
        position={pos}
        popupRoot={popupRoot}
      />
    </>
  );
};

RightClickWrapper.propTypes = {
  children: PropTypes.node.isRequired,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      onClick: PropTypes.func.isRequired,
    })
  ).isRequired,
  popupRoot: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
};
