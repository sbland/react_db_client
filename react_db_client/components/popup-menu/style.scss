// @import 'variables';
$white: white;
$unit: 1rem;
$grey-40: #bbbbbb;
$primaryColour: var(--primary-color, #888);


.popupMenuWrap {
  position: absolute;

  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  z-index: 9999;
}


.popupMenu {
  background: white;
  z-index: 10;

  position: absolute;
  box-shadow: 1px 1px 20px 10px rgba(100, 100, 100, 0.15);
}


.popupMenu_list {
  list-style: none;
  margin: 0;
  border-radius: 5px;
}

.popupMenu_menuItem {
  button {
    width: 100%;
    background: $white;

    padding: ($unit *.5) $unit;

    &:hover {
      background: $grey-40;
    }
  }
}

// .popupMenu_rightClickWrapper {
//   position: absolute;

//   left: 0;
//   right: 0;
//   top: 0;
//   bottom: 0;
// }

.popupMenu_righClickWrap {
  width: 100%;
  height: 100%;

  &.active {
    border:1px $primaryColour dashed;
  }
}

.popupMenu_overlay {
  position: absolute;
  // background: black;
  opacity: 0.4;
  // cursor: pointer;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  // background:url(
  //   data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAcAAAAHCAYAAADEUlfTAAAAMElEQVQImXXJsREAIAwDMS/kZX47b0sFByQpVElAOrajKYCaO0re8eQfJ7sAoimALAAwSVfRiD/CAAAAAElFTkSuQmCC
  // ) repeat;
}
