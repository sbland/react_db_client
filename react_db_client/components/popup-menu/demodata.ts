import { IItem } from './popup-menu';

export const demoItems: IItem[] = [
  { uid: 'a', label: 'Item A', onClick: () => {} },
  { uid: 'b', label: 'Item B', onClick: () => {} },
  { uid: 'c', label: 'Item C', onClick: () => {} },
];

export const demoItemsEmpty = [];
