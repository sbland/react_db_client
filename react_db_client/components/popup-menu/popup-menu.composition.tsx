import React from 'react';
import { demoItems } from './demodata';

import { PopupMenu, RightClickWrapper } from './popup-menu';

const WrapperB = ({ children }) => {
  const handleOnClick = () => {
    alert('TEST');
  };
  return React.cloneElement(children, { ...children.props, onClick: handleOnClick });
};

const App = () => {
  return (
    <WrapperB>
      <div
        style={{
          position: 'relative',
          width: '300px',
          height: '300px',
          background: 'red',
        }}
      />
    </WrapperB>
  );
};

const WrapperC = ({ children }) => {
  const handleOnClick = () => {
    alert('TEST');
  };
  return React.cloneElement(children, { ...children.props, onClick: handleOnClick });
};

const InnerWrap = () => {
  return (
    <div
      style={{
        position: 'relative',
        width: '300px',
        height: '300px',
        background: 'red',
      }}
    />
  );
};

const AppB = () => {
  return (
    <WrapperC>
      <InnerWrap />
    </WrapperC>
  );
};

const Wrapper = () => (
  <div
    style={{
      position: 'relative',
      width: '300px',
      height: '300px',
      background: 'red',
    }}
  />
);
export const defaultPopupMenu = () => (
  <PopupMenu items={demoItems} position={{ x: 0, y: 0 }} onCloseCallback={() => {}} />
);
export const defaultPopupMenuOpen = () => (
  <PopupMenu
    items={demoItems}
    isOpenOverride
    onCloseCallback={() => {}}
    position={{ x: 0, y: 0 }}
  />
);
export const interactivePopupMenu = () => (
  <RightClickWrapper
    items={[
      { uid: 'A', label: 'Item A', onClick: () => {} },
      { uid: 'B', label: 'Item B', onClick: () => {} },
    ]}
  >
    <div
      style={{
        // position: 'relative',
        width: '300px',
        height: '300px',
        background: 'red',
      }}
    />
  </RightClickWrapper>
);
export const interactiveMultiplePopupMenu = () => (
  <div
    style={{
      // position: 'relative',
      width: '800px',
      height: '800px',
      background: 'grey',
    }}
  >
    <RightClickWrapper
      items={[
        { uid: 'A', label: 'Item A', onClick: () => {} },
        { uid: 'B', label: 'Item B', onClick: () => {} },
      ]}
    >
      <div
        style={{
          // position: 'relative',
          width: '100px',
          height: '100px',
          background: 'red',
          position: 'absolute',
          transform: 'translate(100px, 100px)',
        }}
      />
    </RightClickWrapper>
    <RightClickWrapper
      items={[
        { uid: 'C', label: 'Item C', onClick: () => {} },
        { uid: 'D', label: 'Item D', onClick: () => {} },
      ]}
    >
      <div
        style={{
          // position: 'relative',
          width: '100px',
          height: '100px',
          background: 'red',
          position: 'absolute',
          transform: 'translate(400px, 400px)',
        }}
      />
    </RightClickWrapper>
  </div>
);
export const interactiveEmbeddedPopupMenu = () => (
  <RightClickWrapper
    items={[
      { uid: 'A', label: 'Item A', onClick: () => {} },
      { uid: 'B', label: 'Item B', onClick: () => {} },
    ]}
  >
    <Wrapper />
  </RightClickWrapper>
);
export const popupMenu1 = () => <App />;
export const popupMenu2 = () => <AppB />;
