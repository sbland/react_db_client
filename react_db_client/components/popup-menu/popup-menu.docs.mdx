---
description: 'A PopupMenu component.'
labels: ['label1', 'label2', 'label3']
---

import { PopupMenu } from './popup-menu';

## React Component creating a popup

A basic popup component

### Component usage
```js
<PopupMenu text="hello from PopupMenu" />
```

### Using props to customize the text

Modify the text to see it change live:
```js live
<PopupMenu text="hello from PopupMenu" />
```
