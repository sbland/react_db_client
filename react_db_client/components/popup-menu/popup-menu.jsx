import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';

import './style.scss';

const getRoot = (inputRoot) => {
  let root = null;
  if (typeof inputRoot == 'object') root = inputRoot;
  if (typeof inputRoot == 'string') root = document.getElementById(inputRoot);
  if (!root) {
    root = document.createElement('div');
    root.setAttribute('id', inputRoot || '_root');
  }
  return root;
};

const PopupMenuItems = ({ items }) =>
  items.map((item) => (
    <li className="popupMenu_menuItem" key={item.uid}>
      <button type="button" className="button-reset" onClick={item.onClick}>
        {item.label}
      </button>
    </li>
  ));

PopupMenuItems.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
    })
  ).isRequired,
};

/**
 * Popup Menu
 *
 * @param {*} {
 *   isOpenOverride,
 *   items,
 * }
 * @returns
 */
export const PopupMenu = ({ isOpenOverride, onCloseCallback, position, items, popupRoot }) => {
  const [isOpen, setIsOpen] = useState(isOpenOverride);

  const _popupRoot = getRoot(popupRoot);

  useEffect(() => {
    setIsOpen(isOpenOverride);
  }, [isOpenOverride]);

  const handleClose = (e) => {
    e.preventDefault();
    setIsOpen(false);
    onCloseCallback(false);
  };

  if (!_popupRoot) return <div>Missing Popup Root</div>;

  if (isOpen) {
    return ReactDOM.createPortal(
      <div className="popupMenuWrap">
        <div
          className="popupMenu"
          style={{
            transform: `translate(${position.x}px, ${position.y}px)`,
          }}
        >
          <ul className="popupMenu_list">
            <PopupMenuItems items={items} />
          </ul>
        </div>
        {/* eslint-disable-next-line max-len */}
        {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */}
        <div className="popupMenu_overlay" onClick={handleClose} onContextMenu={handleClose} />
      </div>,
      _popupRoot
    );
  }
  return <></>;
};

PopupMenu.propTypes = {
  isOpenOverride: PropTypes.bool,
  onCloseCallback: PropTypes.func.isRequired,
  position: PropTypes.shape({
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
  }).isRequired,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      onClick: PropTypes.func.isRequired,
    })
  ).isRequired,
  popupRoot: PropTypes.any,
};

PopupMenu.defaultProps = {
  isOpenOverride: false,
};

export const RightClickWrapper = ({ children, items, popupRoot }) => {
  const [isOpen, setIsOpen] = useState(false);
  const [pos, setPos] = useState({ x: 0, y: 0 });
  const handleContextMenu = (e) => {
    e.preventDefault();
    setIsOpen(true);
    setPos({ x: e.clientX, y: e.clientY });
  };

  return (
    <>
      {children?.type === 'div' ? (
        React.cloneElement(children, { onContextMenu: handleContextMenu })
      ) : (
        // : React.cloneElement(children, { ref: childRef }, [
        //   <div
        //     className="popupMenu_rightClickWrapper"
        //     onContextMenu={handleContextMenu}
        //     style={{
        //       width: '100%',
        //       height: '100%',
        //     }}
        //   />,
        // ])
        <div
          style={{ width: '100%', height: '100%' }}
          className={`popupMenu_righClickWrap ${isOpen && 'active'}`}
          onContextMenu={handleContextMenu}
        >
          {children}
        </div>
      )}
      <PopupMenu
        isOpenOverride={isOpen}
        onCloseCallback={() => setIsOpen(false)}
        items={items}
        position={pos}
        popupRoot={popupRoot}
      />
    </>
  );
};

RightClickWrapper.propTypes = {
  children: PropTypes.node.isRequired,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      onClick: PropTypes.func.isRequired,
    })
  ).isRequired,
  popupRoot: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
};
