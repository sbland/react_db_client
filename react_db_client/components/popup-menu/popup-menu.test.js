import '@samnbuk/react_db_client.helpers.enzyme-setup';
import React from 'react';
import { shallow, mount } from 'enzyme';

import { RightClickWrapper } from './popup-menu';

describe('ItemList', () => {
  it('renders', () => {
    shallow(
      <RightClickWrapper
        items={[
          { uid: 'A', label: 'Item A', onClick: () => {} },
          { uid: 'B', label: 'Item B', onClick: () => {} },
        ]}
      >
        <div
          id="box"
          style={{
            // position: 'relative',
            width: '100px',
            height: '100px',
            background: 'red',
            position: 'absolute',
            transform: 'translate(400px, 400px)',
          }}
        />
      </RightClickWrapper>
    );
  });
  it('matches snapshot', () => {
    const out = mount(
      <RightClickWrapper
        items={[
          { uid: 'A', label: 'Item A', onClick: () => {} },
          { uid: 'B', label: 'Item B', onClick: () => {} },
        ]}
        popupRoot="root"
      >
        <div
          id="box"
          style={{
            // position: 'relative',
            width: '100px',
            height: '100px',
            background: 'red',
            position: 'absolute',
            transform: 'translate(400px, 400px)',
          }}
        />
      </RightClickWrapper>
    );
    expect(out.debug()).toMatchSnapshot();
  });
});
