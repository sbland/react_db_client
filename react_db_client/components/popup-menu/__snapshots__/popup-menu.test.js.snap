// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemList matches snapshot 1`] = `
"<RightClickWrapper items={{...}}>
  <div id=\\"box\\" style={{...}} onContextMenu={[Function: handleContextMenu]} />
  <PopupMenu isOpenOverride={false} onCloseCallback={[Function: onCloseCallback]} items={{...}} position={{...}} popupRoot=\\"root\\">
    <div>
      Missing Popup Root
    </div>
  </PopupMenu>
</RightClickWrapper>"
`;
